kymatio/scattering2d/backend/numpy_backend.py

```diff
@@ -0,0 +1,188 @@
+        
+        Returns
+        -------
+        res : tensor_like
+            tensor such that its fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            F^{-1}(res)[u1, u2] = F^{-1}(x)[u1 * k, u2 * k]
+
+    """
+    def __call__(self, x, k):
+        y = x.reshape(-1, k, x.shape[1] // k, k, x.shape[2] // k)
+        
+        out = y.mean(axis=(1, 3))
+
+        return out
+
+
+class Modulus(object):
+    """
+        This class implements a modulus transform for complex numbers.
+
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+
+        Parameters
+        ---------
+        x: input complex tensor.
+
+        Returns
+        -------
+        output: a real tensor equal to the modulus of x.
+
+    """
+    def __call__(self, x):
+        norm = np.abs(x)
+        return norm
+
+
+def fft(x, direction='C2C', inverse=False):
+    """
+        Interface with torch FFT routines for 2D signals.
+        
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x, inverse=True)
+        
+        Parameters
+        ----------
+        input : tensor
+            complex input for the FFT
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then an error is raised.
+
+    """
+    if direction == 'C2R':
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if direction == 'C2R':
+        output = np.real(np.fft.ifft2(x)) * x.shape[-1] * x.shape[-2]
+    elif direction == 'C2C':
+        if inverse:
+            output = np.fft.ifft2(x) * x.shape[-1] * x.shape[-2]
+        else:
+            output = np.fft.fft2(x)
+
+    return output
+
+
+def cdgmm(A, B, inplace=False):
+    """
+        Complex pointwise multiplication between (batched) tensor A and tensor B.
+        
+        Parameters
+        ----------
+        A : tensor
+            A is a complex tensor of size (B, C, M, N, 2)
+        B : tensor
+            B is a complex tensor of size (M, N) or real tensor of (M, N)
+        inplace : boolean, optional
+            if set to True, all the operations are performed inplace
+        
+        Returns
+        -------
+        C : tensor
+            output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
+
+    """
+    if B.ndim != 2:
+        raise RuntimeError('The dimension of the second input must be 2.')
+
+    if inplace:
+        return np.multiply(A, B, out=A)
+    else:
+        return A * B
+
+
+def empty_like(x, shape):
+    return np.empty(shape, x.dtype)
+
+def squeeze(x, y, index):
+    y[..., index, :, :] = x.squeeze(-3)
+    return y
```

Comment:
This is exactly the same code in the Torch backend squeeze method... would it be so wrong to do `from .torch_backend import squeeze` on the numpy backend? :P

Reply:
yes

Reply:
Yeah I don't think we should mix these. Imagine a user only has NumPy installed and not Torch, then they should still be able to run the code.

Reply:
yeah!